### PR TITLE
Removes the next_dagrun_create_after reset

### LINKED ISSUE #19901: No scheduling when max_active_runs is 1
### Apache Airflow version

2.2.2 (latest released)

### Operating System

Debian

### Versions of Apache Airflow Providers

_No response_

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### What happened

Since version 2.2 a field `DAG.next_dagrun_create_after` is not calculated when `DAG.max_active_runs` is 1.

### What you expected to happen

https://github.com/apache/airflow/blob/fca2b19a5e0c081ab323479e76551d66ab478d07/airflow/models/dag.py#L2466

If this condition is evaluated when a state is "running" then it is incorrect.

### How to reproduce

Create a DAG with a `schedule_interval` and `max_active_runs=1.`

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #21083: A high value of min_file_process_interval & max_active_runs=1 causes stuck dags
### Apache Airflow version

2.2.2

### What happened

When the value of `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL` is set to 86400, all dags whose `MAX_ACTIVE_RUNS` is set to 1 stop executing & remains stuck forever. If the `MAX_ACTIVE_RUNS` is set to 2 or above, or `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL` is set to a lower value (bw 30-300), dags work just fine. 

### What you expected to happen

These 2 settings should be exclusive to each other and there should be no direct impact of one another. 

### How to reproduce

set AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL to 86400
set MAX_ACTIVE_RUNS to 1 on any dag & observe its execution dates. 

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==1!2.3.0
apache-airflow-providers-cncf-kubernetes==1!2.1.0
apache-airflow-providers-elasticsearch==1!2.1.0
apache-airflow-providers-ftp==1!2.0.1
apache-airflow-providers-google==1!6.1.0
apache-airflow-providers-http==1!2.0.1
apache-airflow-providers-imap==1!2.0.1
apache-airflow-providers-microsoft-azure==1!3.3.0
apache-airflow-providers-mysql==1!2.1.1
apache-airflow-providers-postgres==1!2.3.0
apache-airflow-providers-redis==1!2.0.1
apache-airflow-providers-slack==1!4.1.0
apache-airflow-providers-sqlite==1!2.0.1
apache-airflow-providers-ssh==1!2.3.0

### Deployment

Astronomer

### Deployment details

Deployed on AKS via Astronomer Helm chart. 

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by avkirilishin**
closes: #21083
closes: #19901

The `next_dagrun_create_after` reset causes a problem (details in the issue).
But now we check `_should_update_dag_next_dagruns` before `calculate_dagrun_date_fields` call. So we can remove the reset.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by ashb** [comment]
This needs more thought/a deeper review than I can give right now.

I think by removing that then we might have undone the fix where we introduced this change in the first place

**Comment by avkirilishin** [comment]
> I think by removing that then we might have undone the fix where we introduced this change in the first place

I checked this. Let's reconstruct the sequence of events.

- Initially next_dagrun_create_after resets only in `calculate_dagrun_date_fields` and only with other  "next*" attributes (`next_dagrun`, `next_dagrun_data_interval`). It didn't control max_active_runs.
- #18897: unconditional `_update_dag_next_dagruns` call introduced, now `next_dagrun_create_after` used to control max_active_runs
- #19528: `_should_update_dag_next_dagruns` was added, so we don't call `_update_dag_next_dagruns` without check and we don't need `next_dagrun_create_after = NULL` anymore

File: tests/jobs/test_scheduler_job.py [hunk 1]
```diff
@@ -2937,20 +2975,12 @@ def test_more_runs_are_not_created_when_max_active_runs_is_reached(self, dag_mak
...
         self.scheduler_job.processor_agent = mock.MagicMock(spec=DagFileProcessorAgent)
         session = settings.Session()
-        assert session.query(DagRun).count() == 0
-        dag_models = DagModel.dags_needing_dagruns(session).all()
-        self.scheduler_job._create_dag_runs(dag_models, session)
-        dr = session.query(DagRun).one()
-        dr.state == DagRunState.QUEUED
-        assert session.query(DagRun).count() == 1
-        assert dag_maker.dag_model.next_dagrun_create_after is None
+        self.scheduler_job._do_scheduling(session)
```
Comment (ephraimbuddy) [inline tests/jobs/test_scheduler_job.py]: Using `_do_scheduling` directly is flaky and your code here didn't check that when max_active_runs is reached that more dagruns are not created

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
We need to have a test that creates a dagrun and while the dagrun is still running, the scheduler loops again and should not create more dagrun

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
This works fine in that it doesn't create more runs when max_active_runs is reached but it didn't resolve the linked issue.
Can you work on fixing the linked issue too as I like that `next_dagrun_create_after` is no longer nullified

**Comment by avkirilishin** [comment]
> but it didn't resolve the linked issue

This fix resolves the linked issue. The new test `test_runs_are_created_after_max_active_runs_was_reached` verifies this.

I have to explain how it works:
1. The value of `max_active_runs` has been reached, and the value of `next_dagrun_create_after` has been reset: https://github.com/apache/airflow/blob/ab762a5a8ae147ae33500ee3c7e7a73d25d03ad7/airflow/jobs/scheduler_job.py#L975
2. Due to filtering `next_dagrun_create_after`, `DagModel.dags_needing_dagruns` (referred to in `SchedulerJob._create_dagruns_for_dags`) cannot select Dags to create new runs. Comparison predicates with NULL values always return UNKNOWN (==FALSE in that case).
https://github.com/apache/airflow/blob/ab762a5a8ae147ae33500ee3c7e7a73d25d03ad7/airflow/models/dag.py#L2866-L2876
3. Dag sticks and runs are no longer being created
4. If `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL` is low, the scheduler processes dag's file and calls `Dag.bulk_write_to_db`, which calls `calculate_dagrun_date_fields` and calculates the new `next_dagrun_create_after` value
https://github.com/apache/airflow/blob/ab762a5a8ae147ae33500ee3c7e7a73d25d03ad7/airflow/models/dag.py#L2441-L2444
5. But if `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL` is high, nobody calculates `next_dagrun_create_after`
6. This merge request corrects situation by removing the `next_dagrun_create_after` nullify

---

`next_dagrun_create_after` has been reset:

Screenshot 2022-02-04 at 21 51 39

`min_file_process_interval` is 180. After 3 minutes (16:21:36 - 16:24:37), the scheduler processes dag's file and calculates the new `next_dagrun_create_after` value:

Screenshot 2022-02-04 at 21 56 18

There is a gap in launches:

image

Pay attention to the same `last_parsed_time` and `Queued At` in the last two screenshots.

**Comment by avkirilishin** [comment]
I did load testing today, and it turned out that now the scheduler consumes a lot more CPU, since it constantly selects all the necessary DAGs in the `DagModel.dags_needing_dagruns` function. Maybe I can add a dict with the time of the last check? But it won't work for `AIRFLOW__SCHEDULER__NUM_RUNS = 1`

**Comment by ephraimbuddy** [comment]
@avkirilishin , Can you test with `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL=86400`

**Comment by avkirilishin** [comment]
> Can you test with `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL=86400`

@ephraimbuddy 

image

image

image

image

image

**Comment by ephraimbuddy** [comment]
Hi @avkirilishin, I debugged this issue and it turned out that the problem is not that the `next_dagrun_create_after` is being nullified. The problem is that this line https://github.com/apache/airflow/blob/1d170f899bcc87110e55192517270ec89d511ca8/airflow/jobs/scheduler_job.py#L1060 
is returning a wrong figure. You can verify this. 

I have not been able to pin out why it's returning a wrong figure but If you change this line of code:
https://github.com/apache/airflow/blob/1d170f899bcc87110e55192517270ec89d511ca8/airflow/jobs/scheduler_job.py#L1062
to
```python
if self._should_update_dag_next_dagruns(dag, dag_model, active_runs-1):
```
The problem will be gone.

**Comment by avkirilishin** [comment]
> I have not been able to pin out why it's returning a wrong figure but If you change this line of code:
> The problem will be gone.

I tested your suggestion. The problem has not gone away. DAG sticks after the first launch.

image

image

image

**Comment by ephraimbuddy** [comment]
> I tested your suggestion. The problem has not gone away. DAG sticks after the first launch.

Sorry, I have updated my comment, I gave the wrong permalink. You should change line 1088

**Comment by avkirilishin** [comment]
> > I tested your suggestion. The problem has not gone away. DAG sticks after the first launch.
> 
> Sorry, I have updated my comment, I gave the wrong permalink. You should change line 1088
> 
> https://github.com/apache/airflow/blob/1d170f899bcc87110e55192517270ec89d511ca8/airflow/jobs/scheduler_job.py#L1088

This is a good change, and I think we need to apply it. But only it won't work if one sets the DagRun state manually (Success or Failed).

**Comment by ephraimbuddy** [comment]
> This is a good change, and I think we need to apply it. But only it won't work if one sets the DagRun state manually (Success or Failed).

It worked for me. Can you tests too?

**Comment by avkirilishin** [comment]
> > This is a good change, and I think we need to apply it. But only it won't work if one sets the DagRun state manually (Success or Failed).
> 
> It worked for me. Can you tests too?

With main plus "active_runs-1" or with this branch plus "active_runs-1"?

Main plus "active_runs-1" - the manually changed DagRun state leads to a hang for me (with `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL=86400`).
This branch plus "active_runs-1" - the manually changed DagRun state works fine.

**Comment by ephraimbuddy** [comment]
> With main plus "active_runs-1" or with this branch plus "active_runs-1"?
> 
> Main plus "active_runs-1" - the manually changed DagRun state leads to a hang for me (with `AIRFLOW__SCHEDULER__MIN_FILE_PROCESS_INTERVAL=86400`). This branch plus "active_runs-1" - the manually changed DagRun state works fine.

It still works for me in main, not sure why. I will suggest we implement `active_runs-1` and look at fixing it when the state is manually changed. 
We can also create a boolean column on DagModel and instead of nullifying next_dagrun_create_after, we nullify the column. That way, we don't need to calculate the column again. We would just set the new column when max_active_runs is reached and unset it when dagrun finishes running or are marked manually as failed.
That together with `active_runs-1` at lines 1088 and 1062 will be a better fix

**Comment by ephraimbuddy** [comment]
> But we can (and we do it) calculate the current running DagRun count at any time. Why do we need a new column?
> 
> Maybe this way:
> 
> * implement `active_runs-1`
> * remove nullifying `next_dagrun_create_after`
> * and a little later I will redo the load testing, maybe my previous conclusions were wrong
> 
> ?

I want to avoid https://github.com/apache/airflow/pull/21214/files#diff-62c8e300ee91e0d59f81e0ea5d30834f04db71ae74f2e155a10b51056b00b59bR2874. That query is run a lot, the conditional check and providing a list of IDs to look into won't be efficient in large deployments

**Comment by avkirilishin** [comment]
Is there anything else I can do? And whether to close this merge request or not?

**Comment by ephraimbuddy** [comment]
> Is there anything else I can do? And whether to close this merge request or not?

If you don't mind you can apply the `active_runs-1` change and leave out the addition of new column for now until this issue https://github.com/apache/airflow/issues/16078 is resolved.

**Comment by ephraimbuddy** [comment]
I have added the PR: https://github.com/apache/airflow/pull/21413

**Comment by avkirilishin** [comment]
Can we add `test_runs_are_created_after_max_active_runs_was_reached` to #21413 from this merge request?

**Comment by ephraimbuddy** [comment]
Done

**Pull request closed without merging**